types/coin_test.go

```diff
@@ -421,20 +447,35 @@ func TestCoins(t *testing.T) {
 		{"tree", NewInt(0)},
 		{"mineral", NewInt(1)},
 	}
-	dup := Coins{
+	dup1 := Coins{
 		{"gas", NewInt(1)},
 		{"gas", NewInt(1)},
 		{"mineral", NewInt(1)},
 	}
+	dup2 := Coins{
+		{"GAS", NewInt(1)},
+		{"gAs", NewInt(1)},
+	}
+	dup3 := Coins{
+		{"GAS", NewInt(1)},
+		{"gas", NewInt(1)},
+		{"MINERAL", NewInt(1)},
+		{"mineral", NewInt(1)},
+		{"tree", NewInt(1)},
+		{"TREE", NewInt(1)},
+	}
 	neg := Coins{
 		{"gas", NewInt(-1)},
 		{"mineral", NewInt(1)},
 	}
 
 	assert.True(t, good.IsValid(), "Coins are valid")
-	assert.False(t, mixedCase1.IsValid(), "Coins denoms contain upper case characters")
-	assert.False(t, mixedCase2.IsValid(), "First Coins denoms contain upper case characters")
-	assert.False(t, mixedCase3.IsValid(), "Single denom in Coins contains upper case characters")
+	assert.True(t, goodCaps.IsValid(), "Coins all caps are valid")
```

Comment:
Nit: usually this is some sort of table test, no?

Personally, I like the named table test:

```go
cases := map[string]struct { 
  someTestData string
}{
  "all upper case": { "FOO" },
  "mixed case": {"fOO"},
}

for name, tc := range cases {
  t.Run(name, func(t *testing.T) {
    // whatever you want to test
    assert.True(IsValid(tc.someTestData)
  }
}
```

They are all run separately and each failure has the name from the upper test case map.